task3.py

```diff
@@ -0,0 +1,305 @@
+    :param filename: The desired filename (without the extension) for the downloaded CSV file
+    :return: The path to the downloaded CSV file
+    """
+    response = requests.get(URL)
+
+    csv_file_path = os.path.join(output_folder, f'{filename}.csv')
+
+    with open(csv_file_path, 'w', encoding='utf-8') as csv_file:
+        csv_file.write(response.text)
+
+    return csv_file_path
+
+
+def filter_data(csv_file, gender, rows):
+    """
+    Filters the data from a CSV file based on gender and a specified number of rows
+    :param csv_file: The path to the CSV file to be filtered
+    :param gender: The gender to filter by. If None, no gender filtering will be applied
+    :param rows: The maximum number of rows to include in the filtered data. If None, all rows will be included
+    :return: A list of dictionaries representing the filtered data
+    """
+    with open(csv_file, 'r', encoding='utf-8') as csv_file:
+        reader = csv.DictReader(csv_file)
+        filtered_data = [row for row in reader if (not gender or row['gender'].lower() == gender.lower()) and
+                         (not rows or reader.line_num <= rows)]
+    return filtered_data
+
+
+def add_fields(data):
+    """
+    Adds additional fields to the data
+    :param data: A list of dictionaries representing the data to which additional fields will be added
+    """
+    for index, row in enumerate(data, start=2):
+        row['global_index'] = index
+
+        dob_date = datetime.strptime(row['dob.date'], '%Y-%m-%dT%H:%M:%S.%fZ')
+        row['dob.date'] = dob_date.strftime('%m/%d/%Y')
+
+        reg_date = datetime.strptime(row['registered.date'], '%Y-%m-%dT%H:%M:%S.%fZ')
+        row['registered.date'] = reg_date.strftime('%m-%d-%Y, %H:%M:%S')
+
+        row['name.title'] = TITLE_MAPPING.get(row['name.title'], row['name.title'])
+
+        offset_str = row['location.timezone.offset']
+        offset_hours, offset_minutes = map(int, offset_str.split(':'))
+        offset = timedelta(hours=offset_hours, minutes=offset_minutes)
+
+        timezone = pytz.FixedOffset(offset.total_seconds() // 60)
+        current_time = datetime.now(timezone).strftime('%Y-%m-%d %H:%M:%S')
+
+        row['current_time'] = current_time
+
+
+def create_folder_structure(data):
+    """
+    Creates folder structure based on data
+    :param data: A list of dictionaries representing the data used to create the folder structure
+    :return: A nested dictionary representing the folder structure
+    """
+    folder_structure = {}
+    for row in data:
+        dob_year = datetime.strptime(row['dob.date'], '%m/%d/%Y').year
+        decade = f'{dob_year // 10}0-th'
+        country = row['location.country']
+
+        if decade not in folder_structure:
+            folder_structure.setdefault(decade, {})
+        if country not in folder_structure[decade]:
+            folder_structure[decade].setdefault(country, [])
```

Comment:
you don't need this checks before setdefault.
please work out better how setdefault works.